# STM Transit Delay Data Modeling

## Overview

This notebook explores tree-based regression and classification models in order to find the one that predicts STM transit delays with the best accuracy. The featured models are XGBoost, LightGBM and CatBoost.

## Data Description

`exp_trip_duration`: Expected duration of a trip, in seconds.<br>
`route_direction`: Route direction in degrees.<br>
`route_type_Night`, `route_type_High Frequency` : One-Hot features for types of bus lines<br>
`stop_location_group`: Stop cluster based on coordinates.<br>
`stop_distance`: Distance between the previous and current stop, in meters.<br>
`trip_phase_end`: One-Hot feature for trip progress.<br>
`exp_delay_prev_stop`: Expected duration between the previous and current stop, in seconds.<br>
`wheelchair_boarding`: Indicates if the stop is accessible for people in wheelchair.<br>
`sch_rel_Scheduled`: One-Hot feature for schedule relationship.<br>
`time_of_day_evening`, `time_of_day_morning`, `time_of_day_night`: One-Hot features for time of day.<br>
`is_weekend`: Boolean value if the day of week in on the weekend.<br>
`is_peak_hour`: Boolean value indicating if the sheduled arrival time is at peak hour.<br>
`temperature`: Air temperature at 2 meters above ground, in Celsius.<br>
`relative_humidity`: Relative humidity at 2 meters above ground, in percentage.<br>
`precipitation`: Total precipitation (rain, showers, snow) sum of the preceding hour, in millimeters.<br>
`pressure`: Atmospheric air pressure reduced to mean sea level (msl), in hPa.<br>
`cloud_cover`: Total cloud cover as an area fraction.<br>
`windspeed`: Wind speed at 10 meters above ground, in kilometers per hour.<br>
`wind_direction`: Wind direction at 10 meters above ground.<br>
`delay`: Difference between real and scheduled arrival time, in seconds.<br>
`delay_class`: Delay category, from early to late.

## Imports

In [1]:
from catboost import CatBoostRegressor, CatBoostClassifier
import joblib
import lightgbm as lgb
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pickle
import random
import seaborn as sns
import shap
from sklearn.metrics import cohen_kappa_score, confusion_matrix, f1_score, classification_report, mean_absolute_error, root_mean_squared_error, r2_score
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV, train_test_split
import sys
import xgboost as xgb

In [2]:
# Import custom code
sys.path.insert(0, '..')
from scripts.custom_functions import DELAY_CLASS

In [3]:
df = pd.read_parquet('../data/preprocessed.parquet')
print(f'Shape of dataset: {df.shape}')

Shape of dataset: (7523993, 22)


## Split the data

In [4]:
# Due to the large volume of data, use 60% for training
df_train, df_temp = train_test_split(
    df, 
    test_size=0.4, 
    stratify=df['delay_class'], # stratify with delay class
    random_state=42
)

In [5]:
# Split validation and test sets
df_val, df_test = train_test_split(
  df_temp,
  test_size=0.5,
  stratify=df_temp['delay_class'],
  random_state=42
)

del df_temp

In [6]:
# Check if the delay class distribution is preserved
print(df['delay_class'].value_counts(normalize=True))
print(df_train['delay_class'].value_counts(normalize=True))
print(df_val['delay_class'].value_counts(normalize=True))
print(df_test['delay_class'].value_counts(normalize=True))

delay_class
1    0.884209
2    0.107581
0    0.008210
Name: proportion, dtype: float64
delay_class
1    0.884209
2    0.107581
0    0.008210
Name: proportion, dtype: float64
delay_class
1    0.884209
2    0.107581
0    0.008210
Name: proportion, dtype: float64
delay_class
1    0.884209
2    0.107580
0    0.008210
Name: proportion, dtype: float64


In [7]:
# Separate features from target variables
feature_cols = [col for col in df.columns if col not in ['delay', 'delay_class']]

X_train = df_train[feature_cols]
X_val = df_val[feature_cols]
X_test = df_test[feature_cols]

y_reg_train = df_train['delay']
y_reg_val = df_val['delay']
y_reg_test = df_test['delay']

y_class_train = df_train['delay_class']
y_class_val = df_val['delay_class']
y_class_test = df_test['delay_class']

**Scaling**

Since only tree-based models are explored in this project, scaling is not needed because the models are not sensitive to the absolute scale or distribution of the features.

## Regression Model

### Fit Base Models

All models allow to setup a number of rounds and early stopping. To start, all models will run 100 rounds with an early stopping of 3.

In [8]:
# Create dataframe to track metrics
reg_metrics_df = pd.DataFrame(columns=['model', 'MAE', 'RMSE', 'R²'])

In [9]:
def add_reg_metrics(reg_metrics_df:pd.DataFrame, y_pred:pd.Series, y_val:pd.Series, model_name:str) -> pd.DataFrame:
	mae = mean_absolute_error(y_val, y_pred)
	rmse = root_mean_squared_error(y_val, y_pred)
	r2 = r2_score(y_val, y_pred)

	reg_metrics_df.loc[len(reg_metrics_df)] = [model_name, mae, rmse, r2]
	return reg_metrics_df

#### XGBoost

In [10]:
# Create regression matrices
xg_train_data = xgb.DMatrix(X_train, y_reg_train, enable_categorical=False)
xg_val_data = xgb.DMatrix(X_val, y_reg_val, enable_categorical=False)
xg_eval_set = [(xg_train_data, 'train'), (xg_val_data, 'validation')]

In [11]:
# Train model
xg_reg_base = xgb.train(
  params= {'objective': 'reg:squarederror', 'tree_method': 'hist'},
  dtrain=xg_train_data,
  num_boost_round=100,
  evals=xg_eval_set,
  verbose_eval=10,
  early_stopping_rounds=3
)

[0]	train-rmse:152.42343	validation-rmse:152.79920
[10]	train-rmse:149.66869	validation-rmse:150.04058
[20]	train-rmse:148.86776	validation-rmse:149.26858
[30]	train-rmse:148.22290	validation-rmse:148.63382
[40]	train-rmse:147.69113	validation-rmse:148.14166
[50]	train-rmse:147.26310	validation-rmse:147.73195
[60]	train-rmse:146.87950	validation-rmse:147.36623
[70]	train-rmse:146.48733	validation-rmse:146.99572
[80]	train-rmse:146.25509	validation-rmse:146.78115
[90]	train-rmse:145.90587	validation-rmse:146.45284
[99]	train-rmse:145.62521	validation-rmse:146.19715


In [13]:
# Evaluate model
y_pred = xg_reg_base.predict(xg_val_data)

reg_metrics_df = add_reg_metrics(reg_metrics_df, y_pred, y_reg_val, 'xg_reg_base')
reg_metrics_df

model        MAE        RMSE       R²
0  xg_reg_base  78.730971  146.197149  0.09851

**MAE**<br>
On average, the predictions are off by 79 seconds, which is not very good.

**RMSE**<br>
The higher RMSE compared to MAE suggests that there are some significant prediction errors that influence the overall error metric.

**R²**<br>
The model explains 9.85% of the variance, which indicates the model is a poor fit to the data.

#### LightGBM

In [ ]:
# Train model
lgb_train_data = lgb.Dataset(X_train, label=y_reg_train)
lgb_val_data = lgb.Dataset(X_val, label=y_reg_val, reference=lgb_train_data)

lgb_reg_base = lgb.train(
    {
        'objective': 'regression',
        'metric': 'rmse',
        'learning_rate': 0.05,
        'max_depth': -1
    },
    lgb_train_data,
    valid_sets=[lgb_val_data],
    num_boost_round=100,
    callbacks=[lgb.early_stopping(stopping_rounds=3)]
)

In [ ]:
# Evaluate model
y_pred = lgb_reg_base.predict(X_val)

reg_metrics_df = add_reg_metrics(reg_metrics_df, y_pred, y_reg_val, 'lgb_reg_base')
reg_metrics_df

Overall, the LightGBM model performs worse than XGBoost.

#### CatBoost

In [ ]:
# Fit model
cat_reg_base = CatBoostRegressor(
    iterations=100,
    learning_rate=0.05,
    depth=10,
    random_seed=42,
    verbose=10
)

cat_reg_base.fit(X_train, y_reg_train, eval_set=(X_val, y_reg_val), early_stopping_rounds=3)

In [ ]:
# Evaluate model
y_pred = cat_reg_base.predict(X_val)

reg_metrics_df = add_reg_metrics(reg_metrics_df, y_pred, y_reg_val, 'cat_reg_base')
reg_metrics_df

CatBoost performs slighly better than LightGBM but worse than XGBoost. Without hyperparameter tuning, XGBoost seems to capture a bit more of the underlying patterns than the two other models, but it's still a poor fit for the data.

### Hyperparameter Tuning

Because of the large volume of data, a randomized search will be performed for each model instead of a grid search, wihout K-Fold cross-validation.

#### XGBoost

In [ ]:
n_estimators_options = list(range(100, 700, 100))
max_depth_options = [3, 5, 7]
learning_rate_options = [0.01, 0.1, 0.2]
subsample_options = [0.6, 0.8, 1.0]
colsample_bytree_options = [0.6, 0.8, 1.0]

best_model = None
best_params = {}
best_rmse = float('inf')

for i in range(1, 21): # 20 combinations
  params = {
    'n_estimators': random.choice(n_estimators_options),
    'max_depth': random.choice(max_depth_options),
    'learning_rate': random.choice(learning_rate_options),
    'subsample': random.choice(subsample_options),
    'colsample_bytree': random.choice(colsample_bytree_options)
  }

  print(f'Combination #{i}: {params}')

  xgb_model = xgb.XGBRegressor(
    objective='reg:squarederror',
    n_estimators=params['n_estimators'],
    max_depth=params['max_depth'],
    learning_rate=params['learning_rate'],
    subsample=params['subsample'],
    colsample_bytree=params['colsample_bytree'],
    random_state=42)
  
  xgb_model.fit(X_train, y_reg_train)

  y_pred = xgb_model.predict(X_val)
  rmse = root_mean_squared_error(y_reg_val, y_pred)
  print(f'RMSE: {rmse}\n')

  if rmse < best_rmse:
    best_model = xgb_model
    best_params = params
    best_rmse = rmse

if best_model != None:
  print(f'Best RMSE: {best_rmse}')
  print(f'Best parameters: {best_params}')
else:
  print('No better model was found.')

In [ ]:
# Train best model with more boost rounds
xg_best_params = best_params

xg_reg_tuned = xgb.train(
  params= {
    'objective': 'reg:squarederror',
    'tree_method': 'hist',
    'n_estimators': xg_best_params['n_estimators'],
    'max_depth': xg_best_params['max_depth'],
    'learning_rate': xg_best_params['learning_rate'],
    'subsample': xg_best_params['subsample'],
    'colsample_bytree': xg_best_params['colsample_bytree'],
  },
  dtrain=xg_train_data,
  num_boost_round=1000,
  evals=evals,
  verbose_eval=10,
  early_stopping_rounds=50
)

In [ ]:
# Evaluate model
y_pred = xg_reg_tuned.predict(X_val)

reg_metrics_df = add_reg_metrics(reg_metrics_df, y_pred, y_reg_val, 'xg_reg_tuned')
reg_metrics_df

There's a significant improvement from the base XGBoost model, especially the R-squared.

#### LightGBM

In [ ]:
n_estimators_options = list(range(100, 1100, 100))
learning_rate_options = [0.01, 0.05, 0.1]
max_depth_options = [5, 10, 15]
num_leaves_options = [20, 30, 40]
min_child_samples_options = [10, 20, 30]
subsample_options = [0.8, 1.0]
colsample_bytree_options = [0.8, 1.0]

best_model = None
best_params = {}
best_rmse = float('inf')

for i in range(1, 21):
  params = {
    'n_estimators': random.choice(n_estimators_options),
    'learning_rate': random.choice(learning_rate_options),
    'max_depth': random.choice(max_depth_options),
    'num_leaves': random.choice(num_leaves_options),
    'min_child_samples': random.choice(min_child_samples_options),
    'subsample': random.choice(subsample_options),
    'colsample_bytree': random.choice(colsample_bytree_options)
  }

  print(f'Combination #{i}: {params}')

  lgb_model = lgb.LGBMRegressor(
    n_estimators=params['n_estimators'],
    learning_rate=params['learning_rate'],
    max_depth=params['max_depth'],
    num_leaves=params['num_leaves'],
    min_child_samples=params['min_child_samples'],
    subsample=params['subsample'],
    colsample_bytree=params['colsample_bytree'],
    random_state=42,
    n_jobs=-1,
    verbosity=-1)

  lgb_model.fit(X_train, y_reg_train)

  y_pred = lgb_model.predict(X_val)
  rmse = root_mean_squared_error(y_reg_val, y_pred)
  print(f'RMSE: {rmse}\n')

  if rmse < best_rmse:
    best_model = lgb_model
    best_params = params
    best_rmse = rmse

if best_model != None:
  print(f'Best RMSE: {best_rmse}')
  print(f'Best parameters: {best_params}')
else:
  print('No better model was found.')

In [ ]:
# Train model with more boost rounds and early stopping
lgb_best_params = best_params
lgb_reg_tuned = lgb.train(
    {
        'objective': 'regression',
        'metric': 'rmse',
        'n_estimators': lgb_best_params['n_estimators'],
        'learning_rate': lgb_best_params['learning_rate'],
        'max_depth': lgb_best_params['max_depth'],
        'num_leaves': lgb_best_params['num_leaves'],
        'min_child_samples': lgb_best_params['min_child_samples'],
        'subsample': lgb_best_params['subsample'],
        'colsample_bytree': lgb_best_params['colsample_bytree']
    },
    lgb_train_data,
    valid_sets=[lgb_val_data],
    num_boost_round=1000,
    callbacks=[lgb.early_stopping(stopping_rounds=50)]
)

In [ ]:
# Evaluate model
y_pred = lgb_reg_tuned.predict(X_val)

reg_metrics_df = add_reg_metrics(reg_metrics_df, y_pred, y_reg_val, 'lgb_reg_tuned')
reg_metrics_df

Interpret results

#### CatBoost

In [ ]:
iterations_options = list(range(500, 1100, 100))
learning_rate_options = [0.01, 0.05, 0.1]
depth_options = [6, 8, 10]
l2_leaf_reg_options = [1, 3, 5]
border_count_options = [32, 64, 128]
bagging_temperature_options = [0, 1, 5]

best_model = None
best_params = {}
best_rmse = float('inf')

for i in range(1, 21): # 30 combinations
  params = {
    'iterations': random.choice(iterations_options),
    'learning_rate': random.choice(learning_rate_options),
    'depth': random.choice(depth_options) ,
    'l2_leaf_reg': random.choice(l2_leaf_reg_options),
    'border_count': random.choice(border_count_options),
    'bagging_temperature': random.choice(bagging_temperature_options)
  }

  print(f'Combination #{i}: {params}')

  cat_model = CatBoostRegressor(
    iterations=params['iterations'],
    learning_rate=params['iterations'],
    depth=params['depth'],
    l2_leaf_reg=params['l2_leaf_reg'],
    border_count=params['border_count'],
    bagging_temperature=params['bagging_temperature'],
    verbose=None,
    random_seed=42)

  cat_model.fit(X_train, y_reg_train)

  y_pred = cat_model.predict(X_val)
  rmse = root_mean_squared_error(y_reg_val, y_pred)
  print(f'RMSE: {rmse}\n')

  if rmse < best_rmse:
    best_model = cat_model
    best_params = params
    best_rmse = rmse

if best_model != None:
  print(f'Best RMSE: {best_rmse}')
  print(f'Best parameters: {best_params}')
else:
  print('No better model was found.')

In [ ]:
# Train best model with more iterations
cat_best_params = best_params

cat_reg_tuned = CatBoostRegressor(
    iterations=cat_best_params['iterations'],
    learning_rate=cat_best_params['learning_rate'],
    depth=cat_best_params['depth'],
    l2_leaf_reg=cat_best_params['l2_leaf_reg'],
    border_count=cat_best_params['border_count'],
    bagging_temperature=cat_best_params['bagging_temperature'],
    random_seed=42,
    verbose=50
)

cat_reg_tuned.fit(X_train, y_reg_train, eval_set=(X_val, y_reg_val), early_stopping_rounds=50)

In [ ]:
# Best model
cat_reg_tuned = best_model
cat_best_params = best_params

In [ ]:
# Evaluate model
y_pred = cat_reg_tuned.predict(X_val)

reg_metrics_df = add_reg_metrics(reg_metrics_df, y_pred, y_reg_val, 'cat_reg_tuned')
reg_metrics_df

In [ ]:
# Get model with lowest RMSE
reg_metrics_df.nsmallest(n=1, columns='RMSE')

The best model is XGBoost. This is the model that will be used for the rest of the analysis.

### Residual Analysis

In [ ]:
# Get best model
best_model = xg_reg_tuned
y_pred = best_model.predict(X_val)

In [ ]:
# Plot residuals
fig, (ax1, ax2) = plt.subplots(nrows=1, ncols=2, figsize=(15, 7))

# Predicted vs. actual values
ax1.scatter(x=y_pred, y=y_true)
ax1.set_title('Predicted vs. Actual values')
ax1.set_xlabel('Predicted delay (seconds)')
ax1.set_ylabel('Actual delay (seconds)')
ax1.grid(True)

# Residuals
residuals = y_true - y_pred
ax2.scatter(x=y_pred, y=residuals)
ax2.set_title('Residual Plot')
ax2.set_xlabel('Predicted Delay (seconds)')
ax2.set_ylabel('Residuals (seconds)')
ax2.axhline(0, linestyle='--', color='orange')
ax2.grid(True)

fig.suptitle('Residual Analysis', fontsize=18)
fig.tight_layout()
fig.savefig(f'../images/residual_analysis_{model_name}.png', bbox_inches='tight')
plt.show()

### Feature Importances

#### MDI

#### SHAP Plots

### Feature Pruning

### Retrain Model with Best Features

### Retune Parameters

## Classification Model

### Fit Base Models

In [ ]:
# Create dataframe to track metrics
class_metrics_df = pd.DataFrame(columns=['model', 'params', 'f1_macro', 'kappa'])

In [ ]:
def add_class_metrics(class_metrics_df:pd.DataFrame, model, y_pred:pd.Series, y_val:pd.Series, model_name:str) -> pd.DataFrame:
	f1_macro = f1_score(y_val, y_pred, average='macro')
	kappa = cohen_kappa_score(y_val, y_pred)

	class_metrics_df.loc[len(class_metrics_df)] = [model_name, model.get_params(), f1_macro, kappa]
	return class_metrics_df

#### XGBoost

#### LightGBM

#### CatBoost

### Hyperparameter Tuning

### Feature Importances

### Feature Pruning

### Retrain Model with Best Features

### Retune Parameters

## Final Model

### Evaluate with Test Set

### Make Prediction

In [ ]:
# Save best model
joblib.dump(best_model, 'best_xgb_model.pkl')

## End

# STM Transit Delay Data Modeling Draft

### Random Forest Regressor

#### Fit Model

In [ ]:
# Fit base model
rf_reg_base = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1)
rf_reg_base.fit(X_train_sample, y_reg_train_sample)

#### Evaluate Model

In [ ]:
# Calculate metrics
y_pred = rf_reg_base.predict(X_val)
reg_metrics_df = add_reg_metrics(reg_metrics_df, y_pred, y_reg_val, 'rf_reg_base')
reg_metrics_df

**MAE**<br>
On average, the predictions are off by 74 seconds, which is not very good.

**RMSE**<br>
The higher RMSE compared to MAE suggests that there are some significant prediction errors that influence the overall error metric.

**R²**<br>
The model explains 10.85% of the variance, which indicates the model is a poor fit to the data.

In [ ]:
# Plot residual analysis
plot_residuals(y_pred, y_reg_val, 'rf_reg_base')

Interpret plot

### XGBoost Regressor

#### Fit Base Model

In [ ]:
# Train a model


#### Evaluate Model

In [ ]:
y_pred = xg_reg_base.predict(X_val)

reg_metrics_df = add_reg_metrics(reg_metrics_df, y_pred, y_reg_val, 'xg_reg_base')
reg_metrics_df

**MAE:**

This is a slight improvement over the Random Forest model. The model is now, on average, 73.9 seconds off in its predictions, which is a reduction of almost 10 seconds.

**RMSE:**

The RMSE has also decreased compared to the previous model indicating that the XGBoost model is performing better and has reduced the impact of large errors.

**R²:**

This is a substantial improvement from -4.67%. With an R-squared of 12.52%, the XGBoost model explains more variance in the data, which shows that it's capturing more of the underlying patterns than the previous model. However, it's still a poor fit to the data.

In [ ]:
# Plot residual analysis
plot_residuals(y_pred, y_reg_val, 'xg_reg_base')

#### Hyperparameter tuning

In [ ]:
# Perform GridSearch with 5-fold CV
xgb = xgb.XGBRegressor(objective='reg:squarederror', random_state=42)

param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2],
    'subsample': [0.6, 0.8, 1.0],
    'colsample_bytree': [0.6, 0.8, 1.0]
}

grid_search = GridSearchCV(
    estimator=xgb,
    param_grid=param_grid,
    scoring='neg_mean_absolute_error',
    cv=5,
    n_jobs=-1,
    verbose=1
)

grid_search.fit(X_train, y_reg_train)

In [ ]:
# Best model
xg_reg_tuned = grid_search.best_estimator_
xg_best_params = grid_search.best_params_

In [ ]:
y_pred = xg_reg_tuned.predict(X_val)

reg_metrics_df = add_reg_metrics(reg_metrics_df, y_pred, y_reg_val, 'xg_reg_tuned')
reg_metrics_df

### LightGBM Regressor

#### Fit Base Model

#### Evaluate Model

In [ ]:
y_pred = lgb_reg_base.predict(X_val)

reg_metrics_df = add_reg_metrics(reg_metrics_df, y_pred, y_reg_val, 'lgb_reg_base')
reg_metrics_df

#### Hyperparameter tuning

In [ ]:
lgb_param_grid = {
    'n_estimators': [100, 500, 1000],
    'learning_rate': [0.01, 0.05, 0.1],
    'max_depth': [5, 10, 15],
    'num_leaves': [20, 31, 40],
    'min_child_samples': [10, 20, 30],
    'subsample': [0.8, 1.0],
    'colsample_bytree': [0.8, 1.0]
}

lgb_model = lgb.LGBMRegressor(random_state=42)

lgb_grid = GridSearchCV(
    estimator=lgb_model,
    param_grid=lgb_param_grid,
    cv=3,  # 3-fold cross-validation
    scoring='neg_mean_squared_error',
    verbose=1,
    n_jobs=-1
)

lgb_grid.fit(X_train, y_reg_train)

In [ ]:
lgb_reg_tuned = lgb_grid.best_estimator_

In [ ]:
# Best parameters and score
print("Best Parameters for LightGBM:")
print(lgb_grid.best_params_)
print(f"Best RMSE: {-lgb_grid.best_score_ ** 0.5}")

In [ ]:
y_pred = lgb_reg_tuned.predict(X_val)

reg_metrics_df = add_reg_metrics(reg_metrics_df, y_pred, y_reg_val, 'lgb_reg_tuned')
reg_metrics_df

### CatBoost Regressor

In [ ]:
cat_reg_base = CatBoostRegressor(
    iterations=1000,
    learning_rate=0.05,
    depth=10,
    random_seed=42,
    verbose=100
)

cat_reg_base.fit(X_train, y_reg_train, eval_set=(X_val, y_reg_val), early_stopping_rounds=50)

In [ ]:
y_pred = cat_reg_base.predict(X_val)

reg_metrics_df = add_reg_metrics(reg_metrics_df, y_pred, y_reg_val, 'cat_reg_base')
reg_metrics_df

## Classification Model

### Random Forest Classifier

#### Fit Base Model

In [ ]:
# class_weight helps for rare classes
rf_class_base = RandomForestClassifier(n_estimators=100, random_state=42, class_weight='balanced', n_jobs=-1) 
rf_class_base.fit(X_train, y_class_train)

#### Evaluate Model

In [ ]:
y_pred = rf_class_base.predict(X_val)

class_metrics_df = add_class_metrics(class_metrics_df, rf_class_base, y_pred, y_class_val, 'rf_class_base')
class_metrics_df[['model', 'f1_macro', 'kappa']]

Interpret results

In [ ]:
class_labels = DELAY_CLASS.values()

In [ ]:
def print_class_report(y_val, y_pred, labels):
	print(classification_report(y_val, y_pred, target_names=labels))

In [ ]:
# Classification report
print_class_report(y_class_val, y_pred, class_labels)

In [ ]:
def plot_confusion_matrix(y_val, y_pred, labels:list, model_name:str):
	cm = confusion_matrix(y_val, y_pred)
	plt.figure(figsize=(8, 6))
	sns.heatmap(cm, annot=True, fmt='d', cmap='crest', xticklabels=labels, yticklabels=labels)
	plt.xlabel('Predicted', fontsize=14)
	plt.ylabel('Actual', fontsize=14)
	plt.title('Confusion Matrix', fontsize=18)
	plt.tight_layout()
	plt.savefig(f'../images/cm_{model_name}.png', bbox_inches='tight')
	plt.show()

In [ ]:
# Confusion matrix heatmap
plot_confusion_matrix(y_class_val, y_pred, class_labels, 'rf_class_base')

Interpret results

#### Hyperparameter Tuning

In [ ]:
# Define parameter grid
param_dist = {
    'n_estimators': list(range(100, 700, 100)),
    'max_depth': [None, 10, 20, 30, 50],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['sqrt', 'log2'],
    'bootstrap': [True, False],
    'class_weight': ['balanced', 'balanced_subsample']
}

# Initialize base model
rf = RandomForestClassifier(random_state=42, n_jobs=-1)

# Randomized search with 2-fold CV (to save computation time)
random_search = RandomizedSearchCV(
    estimator=rf,
    param_distributions=param_dist,
    n_iter=10, # 10 combinations
    cv=2,
    verbose=1,
    scoring='f1_macro', # optimize for macro F1
    n_jobs=-1
)

# Fit search
random_search.fit(X_train, y_class_train)

In [ ]:
# Best model
rf_class_tuned = random_search.best_estimator_
rf_best_params = random_search.best_params_

In [ ]:
y_pred = rf_class_tuned.predict(X_val)

class_metrics_df = add_class_metrics(class_metrics_df, rf_class_tuned, y_pred, y_class_val, 'rf_class_tuned')
class_metrics_df[['model', 'f1_macro', 'kappa']]

Interpret results

In [ ]:
# Classification report
print_class_report(y_class_val, y_pred, class_labels)

In [ ]:
# Confusion matrix
plot_confusion_matrix(y_class_val, y_pred, class_labels, 'rf_class_tuned')

Interpret results

### XGBoost Classifier

#### Fit Base Model

In [ ]:
xg_class_base = xgb.XGBClassifier(
	objective='multi:softmax',
  	num_class=3,
    eval_metric='mlogloss',
    random_state=42,
)

xg_class_base.fit(X_train, y_class_train)

#### Evaluate Model

In [ ]:
y_pred = xg_class_base.predict(X_val)

class_metrics_df = add_class_metrics(class_metrics_df, xg_class_base, y_pred, y_class_val, 'xg_class_base')
class_metrics_df[['model', 'f1_macro', 'kappa']]

Interpret results

In [ ]:
# Classification report
print_class_report(y_class_val, y_pred, class_labels)

In [ ]:
# Confusion matrix heatmap
plot_confusion_matrix(y_class_val, y_pred, class_labels, 'xg_class_base')

#### Hyperparameter Tuning

In [ ]:
# Define param grid
param_grid = {
    'n_estimators': [100, 200, 400],
    'max_depth': [3, 5, 7, 10],
    'learning_rate': [0.01, 0.05, 0.1, 0.2],
    'subsample': [0.6, 0.8, 1.0],
    'colsample_bytree': [0.6, 0.8, 1.0],
    'min_child_weight': [1, 3, 5],
    'gamma': [0, 0.1, 0.3, 0.5]
}

# Initialize model
xgb = xgb.XGBClassifier(
    objective='multi:softmax',
    num_class=3,
    eval_metric='mlogloss',
    random_state=42
)

# Grid search with 3-fold CV
grid_search = GridSearchCV(
    estimator=xgb,
    param_grid=param_grid,
    cv=3,
    verbose=1,
    scoring='f1_macro',
    n_jobs=-1
)

# Fit
grid_search.fit(X_train, y_class_train)

In [ ]:
# Best model
xg_class_tuned = grid_search.best_estimator_
xg_best_params = grid_search.best_params_

In [ ]:
# Calculate metrics
y_pred = rf_class_tuned.predict(X_val)

class_metrics_df = add_class_metrics(class_metrics_df, rf_class_tuned, y_pred, y_class_val, 'rf_class_tuned')
class_metrics_df[['model', 'f1_macro', 'kappa']]

Interpret results

In [ ]:
# Classification report
print_class_report(y_class_val, y_pred, class_labels)

In [ ]:
# Confusion matrix heatmap
plot_confusion_matrix(y_class_val, y_pred, class_labels, 'rf_class_tuned')

Interpret results

## End